software/firmware/europi.py

```diff
@@ -1,249 +1,278 @@
+    def read_position(self, steps=100, samples=None):
+        """Returns the position as a value between zero and provided integer."""
+        return self.range(steps, samples)
 
-    def handler(self, func): #Allows the function that is run when triggered to be changed
+
+class DigitalReader:
+    def __init__(self, pin, debounce_delay=500):
+        self.pin = Pin(pin, Pin.IN)
+        self.debounce_delay = debounce_delay
+        self.last_pressed = 0
+
+    def value(self):
+        """The current binary value, HIGH (1) or LOW (0)."""
+        # Both the digital input and buttons are normally high, and 'pulled'
+        # low when on, so this is flipped to be more intuitive (1 when on, 0
+        # when off)
+        return 1 - self.pin.value()
+
+    def handler(self, func):
+        """Define the callback func to call when rising edge detected."""
         def bounce_wrapper(pin):
-            if (ticks_ms() - self.last_pressed) > self.debounce_delay: #As long as the debounce time has been reached
-                self.last_pressed = ticks_ms() #Reset the debounce counter to the current time
-                func() #Run the chosen function
-        self.pin.irq(trigger=Pin.IRQ_FALLING, handler=bounce_wrapper) 
-        
+            if (ticks_ms() - self.last_pressed) > self.debounce_delay:
+                self.last_pressed = ticks_ms()
+                func()
+        # Both the digital input and buttons are normally high, and 'pulled'
+        # low when on, so here we use IRQ_FALLING to detect rising edge.
+        self.pin.irq(trigger=Pin.IRQ_FALLING, handler=bounce_wrapper)
+
     def reset_handler(self):
         self.pin.irq(trigger=Pin.IRQ_FALLING)
 
 
+class DigitalInput(DigitalReader):
+    def __init__(self, pin, debounce_delay=0):
+        super().__init__(pin, debounce_delay)
 
 
-#Define all the I/O using the appropriate class and with the pins used
-oled = Display(0,1,OLED_WIDTH,OLED_HEIGHT)
+class Button(DigitalReader):
+    def __init__(self, pin, debounce_delay=200):
+        super().__init__(pin, debounce_delay)
 
-k1 = Knob(27)
-k2 = Knob(28)
 
-b1 = DigitalInput(4)
-b2 = DigitalInput(5)
+class Display(SSD1306_I2C):
+    def __init__(self, sda, scl, width=OLED_WIDTH, height=OLED_HEIGHT, channel=I2C_CHANNEL, freq=I2C_FREQUENCY):
+        i2c = I2C(channel, sda=Pin(sda), scl=Pin(scl), freq=freq)
+        self.width = width
+        self.height = height
+
+        if len(i2c.scan()) == 0:
+            raise Exception(
+                "EuroPi Hardware Error:\nMake sure the OLED display is connected correctly")
+
+        super().__init__(self.width, self.height, i2c)
+
+    def clear(self):
+        """Clear the display upon call."""
+        self.fill(0)
+        self.show()
+
+    def centre_text(self, text):
+        """Split the provided text across 3 lines of display."""

```

Comment:
The '3' here contradicts the '4' below. 

That said 4 sounds real crowded. I'll have to run it and see what it looks like, but my gut says that it may be a better user experience to limit this convenience method to three lines.

Reply:
I added "but we limit it to 3 lines for readability." to the end of the inline comment to explain the discrepancy.